django/apps/registry.py

```diff
@@ -69,53 +70,58 @@ def populate(self, installed_apps=None):
         # populate() might be called by two threads in parallel on servers
         # that create threads before initializing the WSGI callable.
         with self._lock:
-            if self.ready:
-                return
-
-            # app_config should be pristine, otherwise the code below won't
-            # guarantee that the order matches the order in INSTALLED_APPS.
-            if self.app_configs:
-                raise RuntimeError("populate() isn't reentrant")
-
-            # Phase 1: initialize app configs and import app modules.
-            for entry in installed_apps:
-                if isinstance(entry, AppConfig):
-                    app_config = entry
-                else:
-                    app_config = AppConfig.create(entry)
-                if app_config.label in self.app_configs:
+            try:
+                if self.ready:
+                    return
+
+                # If populate is called concurrently, the code below won't
+                # guarantee that the app_config order matches the order in
+                # INSTALLED_APPS.
+                if self._populate_running:
+                    raise RuntimeError("populate() isn't reentrant")
+                self._populate_running = True
```

Comment:
Non-obviously, there's no race condition here because the RLock only allows one thread to get there.

Consider explaining that in a comment, because it looks like a non-atomic compare-and-set.